rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,777 @@
+* FIPS 186-5 approves Ed25519
+* Go BoringCrypto does not support Ed25519
+  <https://github.com/golang/go/blob/cd6676126b7e663e6202e98e2f235fff20d5e858/src/crypto/tls/boring.go#L78-L90>
+* Ed25519 is the only EdDSA curve supported in the Go standard library.
+
+#### Algorithms Summary
+
+* We are probably forced to continue unconditionally using RSA for database
+  certs, I'm assuming this would apply to both client and CA.
+* Ed25519 is a modern favourite for SSH, but TLS (and HSM, KMS) support is lacking.
+* Teleport CAs use separate keypairs for SSH and TLS, they do not need to use
+  the same algorithm.
+* Teleport derives client SSH and TLS certs from the same client keypair,
+  supporting different algorithms for each will require larger changes.
+* It seems it is time to split client SSH and TLS keys to support the popular
+  and secure Ed25519 algorithm for SSH and the widely-suported
+  `ECDSA_P256_SHA256` algorithm for TLS. This will also allows to evolve the
+  algorithms used for each protocol independently in the future.
+
+### UX
+
+The algorithm suite will be configurable in `teleport.yaml` and
+`cluster_auth_preference` as described above.
+
+We will add visibility of the current and configured CA algorithms to `tctl status`.
+
+```
+$ tctl status
+Cluster      cluster-one
+Version      14.0.0-dev
+
+Host CA pins:
+sha256:d2c825ede608da97891b3eaf1dac27e7e1085be5ff71e48a97ff8eb53360e362
+
+Host CA
+expires:         Jun 21 2033 00:03:19 UTC
+last rotated:    Jun 21 2023 00:03:19 UTC
+rotation state:  standby
+SSH algorithm:   Ed25519
+TLS algorithm:   ECDSA_P256_SHA256
+
+User CA
+expires:         Jun 21 2033 00:03:19 UTC
+last rotated:    never
+rotation state:  standby
+SSH algorithm:   RSA2048_PKCS1_SHA512 (recommended-v1 algorithm Ed25519 will take effect during next manual CA rotation)
+TLS algorithm:   RSA2048_PKCS1_SHA256 (recommended-v1 algorithm ECDSA_P256_SHA256 will take effect during next manual CA rotation)
+
+Database CA
+expires:         Jun 21 2033 00:03:19 UTC
+last rotated:    never
+rotation state:  rotating clients (mode: manual, started: Jun 30 2023 00:03:19 UTC, ending: Jul 1 2023 06:03:19 UTC)
+TLS algorithm:   RSA3072_PKCS1_SHA256
+
+...
+```
+
+`tctl auth rotate` will inform users when a rotation is going to trigger an
+algorithm change.
+
+```
+$ tctl auth rotate --manual --type user --phase init
+INFO: Rotation will update the key types for this CA to match the recommended-v1 suite:
+Protocol  Before                After
+SSH       RSA2048_PKCS1_SHA512  Ed25519
+TLS       RSA2048_PKCS1_SHA256  ECDSA_P256_SHA256
+
+Updated rotation phase to "init". To check status use 'tctl status'
+```
+
+### Proto Specification
+
+TODO:
+
+Include any `.proto` changes or additions that are necessary for your design.
+
+### Audit Events
+
+User login and existing certificate generation events will be supplemented with
+info on the algorithms used by the subject and the CA.
+
+### Observability
```

Comment:
Some additional conditions we may want to log:
* When the configuration has been updated but the certs have not been rotated yet (thus the cipher suite is not fully active)
* Any time the suite is overridden (mentioned above)
* Not for MVP, but long term it would be great to be able to propose cipher suite upgrades

Reply:
> When the configuration has been updated but the certs have not been rotated yet (thus the cipher suite is not fully active)

Yes we should log this, I will add to the doc. I mention in the UX section this will be visible in `tctl status` also.

> Not for MVP, but long term it would be great to be able to propose cipher suite upgrades

It's easy to put something like this in a log on startup and/or in `tctl status`